### PR TITLE
Standardize package names: lower-case, not Mixed_CASE

### PR DESCRIPTION
**Pull request by citibeth**
@adamjstewart 
This PR converts all package names to lower case.  Benefits:
 1. Conforms to previously-discussed Spack standard.
 2. Make it easier for users to "guess" the right name of a package.  (Is it "Python" or "python"?  "R" or "r"?)
  3. Works better on macOS

### CODE REVIEW

**Comment by adamjstewart** [comment]
Can you elaborate on point 3: "Works better on macOS"?

**Comment by citibeth** [comment]
> Can you elaborate on point 3: "Works better on macOS"?

Running things like `git mv R r` can produce strange results on macOS.  Also, macOS will have problems if we were to create multiple packages that differ only in their case.

**Comment by hartzell** [comment]
The mac file system, by default, is  "case-insensitive but case-preserving", it remembers how you created a file but doesn't use the info for anything. The Google has more info, [e.g. here](http://apple.stackexchange.com/questions/22297/is-bash-in-osx-case-insensitive).

I learned this back in the day when I was working with a project in svn that had both a `makefile` and `Makefile`, checking them out on an OS X machine led to much hilarity.

**Comment by adamjstewart** [comment]
By the way, I made a few changes to how `spack create` chooses a default name in #2351 that should help make things more uniform. @citibeth If you haven't already, can you convert packages like `the_silver_searcher` to `the-silver-searcher`? It goes along the same lines of not knowing whether it is uppercase or lowercase.

**Comment by alalazo** [comment]
@glennpj @JavierCVilla Pinging the two guys that contributed most to `r` packaging.

**Comment by citibeth** [comment]
I concur.  But remember, we use lapack not LAPACK.  x not X.  And so on for
so many other acronym packages.  The only difference is r is only one
letter.

**Comment by tgamblin** [comment]
How strongly do people feel about this?  I would be ok with merging this right before the `0.10` release, as this just brings all the packages inline with the convention we're using for `spack create`.  I wonder if this is a big disruption for R users, though.

@davydden @alalazo @glennpj

**Comment by davydden** [comment]
>  I wonder if this is a big disruption for R users, though.

i never used `R`, but i like the fact that everything is named consistently and there is no difference between `py-foo` and `r-foo`.

**Comment by tgamblin** [comment]
I like the consistency.  Mostly looking for @glennpj's feedback as I believe he has R users.

**Comment by glennpj** [comment]
@tgamblin That change should be fine.

**Comment by alalazo** [comment]
@tgamblin also fine with me

**Comment by tgamblin** [comment]
@citibeth Ok people are ok with this.  Let's get this merged quickly, as it is a lot of changes.  Can you rebase on current `develop` and make sure there are no more mixed-case packages?

**Comment by citibeth** [comment]
OK, done.

**Comment by tgamblin** [comment]
I added a test to enforce the convention, so Travis will fail if new, nonconforming packages appear.

While I was doing that, I encountered an issue.  The updates in this branch do NOT naturally take effect for Mac OS X users, because git doesn't correct the directory names.  This causes tests like `package_sanity` to fail.  It's similar to the issue mentioned above where `mv` works in funny ways on Mac OS X.  This *may* be because I use GNU `mv` from `coreutils` and not the builtin Mac OS X `mv` but I don't know what `git` uses to change names on checkout.

How do we make this not immediately break for Mac OS X users?  It would be nice if we could easily check the branch out and switch to a new one without getting case errors, but I dunno the best way to do it without calling `git ls-tree` and correcting the `builtin` package repo.

@davydden: can you check this branch out without getting test errors on your mac due to uncorrected directory names in `var/spack/repos/builtin/packages`?

**Comment by citibeth** [comment]
Yes you are right, and I don't know the solution.  I discovered that issue
early on when trying to APPLY these fixes on the Mac.  To me that's just
one more reason to try to get this merged.

**Comment by JavierCVilla** [comment]
@citibeth I'm not sure about this but just in case, shouldn't affect the change to lower-case also to these parts of the current R/package.py file?

```
    @property
    def r_lib_dir(self):
        return join_path('rlib', 'R', 'library')
```
```    def setup_environment(self, spack_env, run_env):
        run_env.prepend_path('LIBRARY_PATH',
                             join_path(self.prefix, 'rlib', 'R', 'lib'))
        run_env.prepend_path('LD_LIBRARY_PATH',
                             join_path(self.prefix, 'rlib', 'R', 'lib'))
        run_env.prepend_path('CPATH',
                             join_path(self.prefix, 'rlib', 'R', 'include'))
```
```
        # R extension builds can have a global R executable function
        module.R = Executable(join_path(self.spec.prefix.bin, 'R'))
```
I mean in the `join_path` where R is still appended in upper-case.

**Comment by citibeth** [comment]
No, I don't think it changes that.  The R build creates a directory in
upper-case called 'R'.  That's not something we can (easily) change, nor do
I believe we should try.

**Comment by adamjstewart** [comment]
@citibeth You'll also need to update [this line](https://github.com/LLNL/spack/blob/develop/lib/spack/spack/cmd/create.py#L218) so that `spack create` uses lowercase R.

**Comment by adamjstewart** [comment]
There's also a lot of stuff in `lib/spack/spack/util/naming.py` (and possibly elsewhere) that allows packages to contain `_` and `A-Z` values. If we really want to be thorough, changing these would completely disallow future packages from containing these characters.

**Comment by citibeth** [comment]
That is a decent idea, but out of the purview of what I originally set out
to do.  We can also prevent these things from creeping in by asking for
changes on PRs that have them.

**Comment by adamjstewart** [comment]
That just adds work for us to keep an eye out for that. At some point, a non-conforming package will slip in, and we will end up in the same mess of not being able to lowercase packages on macOS. Better to do it right the first time if you ask me.

**Comment by citibeth** [comment]
Can you do it?

**Comment by adamjstewart** [comment]
Not at the moment (probably not any time soon either).

**Comment by tgamblin** [comment]
I'll try to put in a fix for the directory names.

**Comment by citibeth** [comment]
@tgamblin Is this ready to merge, now that #2562 is in?

**Comment by adamjstewart** [comment]
@citibeth You also need to downcase [this line](https://github.com/LLNL/spack/blob/develop/lib/spack/spack/cmd/create.py#L218) so that `spack create` uses lowercase R.

**Comment by citibeth** [comment]
OK... I found a couple more places that seemed to want it too.  Can you
please tell me if my changes broke something?

**Comment by citibeth** [comment]
@alalazo @tgamblin Are you able to help with the test failures I'm getting here?  It's only failing on Python 2.6.

```
______________ PackageSanityTest.test_all_versions_are_lowercase _______________

self = <spack.test.package_sanity.PackageSanityTest testMethod=test_all_versions_are_lowercase>

    def test_all_versions_are_lowercase(self):

        """Spack package names must be lowercase, and use `-` instead of `_`.

            """

        errors = []

        for name in spack.repo.all_package_names():

            if re.search(r'[_A-Z]', name):

                errors.append(name)

    

>       self.assertEqual([], errors)

E       AssertionError: [] != ['direct_mpich', 'indirect_mpich', 'simple_inheritance', 'trivial_install_test_package']

../../../lib/spack/spack/test/package_sanity.py:70: AssertionError

=============== 1 failed, 481 passed, 1 xfailed in 75.12 seconds ===============
```

**Comment by alalazo** [comment]
@citibeth I don't know if I'll have time later in the day to check the causes, but it seems it's failing because it checks `builtin.mock` instead of `builtin`.

**Comment by citibeth** [comment]
@adamjstewart  I found the problem... unit tests were still using underscores in package names instead of dashes.  But this yields a few questions:

1. The tests that tripped me up were checking for underscores and capital letters in the test repo.  That's all fine, but... are they ALSO doing this check in the main repo?

2. It would still be nice to know why this is failing on Python 2.6 and Mac, but succeeding on other systems.  My best guess is that in some cases, the unit test is checking the main repo; and in others, it is checking the test repo.  That would seem to be wrong, are you able to look into that?

https://travis-ci.org/LLNL/spack/builds/187805917

File: lib/spack/docs/module_file_support.rst [hunk 1]
```diff
@@ -367,14 +367,14 @@ and has similar effects on module file of dependees. Even in this case
 
   with the following snippet:
 
-  .. literalinclude:: ../../../var/spack/repos/builtin/packages/R/package.py
+  .. literalinclude:: ../../../var/spack/repos/builtin/packages/r/package.py
      :pyobject: R.setup_environment
 
   The ``R`` package also knows which environment variable should be modified
```
Comment (adamjstewart) [inline lib/spack/docs/module_file_support.rst]: @citibeth This should be `r`. You might want to look through the documentation to make sure there aren't other capitalized packages being reference. The documentation tests aren't foolproof.

**Comment by citibeth** [comment]
OK, I just fixed up that section of the docs.  I agree with @adamjstewart,
let's get this merged and deal with docs later.  I expect there will be
significant editing/rewriting as we integrate build-system-specific
`Package` superclasses into the docs, I think that's the time to really
edit this stuff carefully.

File: lib/spack/docs/module_file_support.rst [hunk 2]
```diff
@@ -367,14 +367,14 @@ and has similar effects on module file of dependees. Even in this case
 
   with the following snippet:
 
-  .. literalinclude:: ../../../var/spack/repos/builtin/packages/R/package.py
-     :pyobject: R.setup_environment
+  .. literalinclude:: ../../../var/spack/repos/builtin/packages/r/package.py
+     :pyobject: r.setup_environment
```
Comment (adamjstewart) [inline lib/spack/docs/module_file_support.rst]: This isn't going to work. The class is still named `R`, just like `python`'s class is named `Python`. You should remove this particular change.

**Comment by citibeth** [comment]
@tgamblin Can we please get this merged?  It's been approved and re-approved.  I just rebased again...

**Pull request merged**